diff --git a/cassandra/connection.py b/cassandra/connection.py
index ba08ae29..93cf0448 100644
--- a/cassandra/connection.py
+++ b/cassandra/connection.py
@@ -877,6 +877,7 @@ class Connection(object):
     def process_msg(self, header, body):
         self.msg_received = True
         stream_id = header.stream
+        print(f"stream_id = {stream_id!r}")
         if stream_id < 0:
             callback = None
             decoder = ProtocolHandler.decode_message
